# Question 2

Objective:
I can do a text summary that will let me undertsand the important words used in the dataset.
I can first understand about the dataset.
Clean the tweets by removing 'Mentions','Special Characters','HTML','UTF-8-bom' if any.
After which techniques like bag of words that uses stemming and lemmetization can be done to finally get the text summary


# Question 3

In [1]:
import pandas as pd  
import numpy as np
cols = ['sentiment','id','date','query_string','user','text']
df = pd.read_csv("N:/Semester 3/Text Mining/trainingandtestdata/testdata.manual.2009.06.14.csv",header=None, names=cols)

3.a) Display the Columns

In [2]:
df.head()

sentiment  id                          date query_string      user  \
0          4   3  Mon May 11 03:17:40 UTC 2009      kindle2    tpryan   
1          4   4  Mon May 11 03:18:03 UTC 2009      kindle2    vcu451   
2          4   5  Mon May 11 03:18:54 UTC 2009      kindle2    chadfu   
3          4   6  Mon May 11 03:19:04 UTC 2009      kindle2     SIX15   
4          4   7  Mon May 11 03:21:41 UTC 2009      kindle2  yamarama   

                                                text  
0  @stellargirl I loooooooovvvvvveee my Kindle2. ...  
1  Reading my kindle2...  Love it... Lee childs i...  
2  Ok, first assesment of the #kindle2 ...it fuck...  
3  @kenburbary You'll love your Kindle2. I've had...  
4  @mikefish  Fair enough. But i have the Kindle2...

3.b) Dimension of the Dataset

In [5]:
np.shape(df)

(498, 6)

It has 498 observations and 6 Columns

3.d) Datatypes of the columns

In [6]:
print(df.dtypes)

sentiment        int64
id               int64
date            object
query_string    object
user            object
text            object
dtype: object


4. Filter the columns and get only the columns used for anlysis - PreProcess the Data

In [8]:
df.drop(['id','date','query_string','user'],axis=1,inplace=True)

In [9]:
df.head()

sentiment                                               text
0          4  @stellargirl I loooooooovvvvvveee my Kindle2. ...
1          4  Reading my kindle2...  Love it... Lee childs i...
2          4  Ok, first assesment of the #kindle2 ...it fuck...
3          4  @kenburbary You'll love your Kindle2. I've had...
4          4  @mikefish  Fair enough. But i have the Kindle2...

Unwamted coloumns are removed. Now we clean the text.

Checking for tweets for more than 280 words

In [11]:
df['pre_clean_len'] = [len(t) for t in df.text]

In [12]:
df[df.pre_clean_len > 280].head(10)

Empty DataFrame
Columns: [sentiment, text, pre_clean_len]
Index: []

We dont have any. So all the texts are legitimate

In [13]:
import sys
!{sys.executable} -m pip install sklearn

  Created wheel for sklearn: filename=sklearn-0.0-py2.py3-none-any.whl size=1316 sha256=612ca16d2dc10bee4080d0efcffd502bcf07aadb8375a3bd78cf4627be7090d7
  Stored in directory: c:\users\amitniki\appdata\local\pip\cache\wheels\22\0b\40\fd3f795caaa1fb4c6cb738bc1f56100be1e57da95849bfc897
Successfully built sklearn


Trying to remove unnecessary symbols like '@', removing htmls, and others

In [20]:
from bs4 import BeautifulSoup
import re
from nltk.tokenize import WordPunctTokenizer
tok = WordPunctTokenizer()
pat1 = r'@[A-Za-z0-9]+'
pat2 = r'https?://[A-Za-z0-9./]+'
combined_pat = r'|'.join((pat1, pat2))
def tweet_cleaner(text):
    soup = BeautifulSoup(text, 'lxml')
    souped = soup.get_text()
    stripped = re.sub(combined_pat, '', souped)
    try:
        clean = stripped.decode("utf-8-sig").replace(u"\ufffd", "?")
    except:
        clean = stripped
    letters_only = re.sub("[^a-zA-Z]", " ", clean)
    lower_case = letters_only.lower()
    # During the letters_only process two lines above, it has created unnecessay white spaces,
    # I will tokenize and join together to remove unneccessary white spaces
    words = tok.tokenize(lower_case)
    return (" ".join(words)).strip()
testing = df.text[:100]
test_result = []
for t in testing:
    test_result.append(tweet_cleaner(t))
test_result

['i loooooooovvvvvveee my kindle not that the dx is cool but the is fantastic in its own right',
 'reading my kindle love it lee childs is good read',
 'ok first assesment of the kindle it fucking rocks',
 'you ll love your kindle i ve had mine for a few months and never looked back the new big one is huge no need for remorse',
 'fair enough but i have the kindle and i think it s perfect',
 'no it is too big i m quite happy with the kindle',
 'fuck this economy i hate aig and their non loan given asses',
 'jquery is my new best friend',
 'loves twitter',
 'how can you not love obama he makes jokes about himself',
 'check this video out president obama at the white house correspondents dinner',
 'i firmly believe that obama pelosi have zero desire to be civil it s a charade and a slogan but they want to destroy conservatism',
 'house correspondents dinner was last night whoopi barbara sherri went obama got a standing ovation',
 'watchin espn jus seen this new nike commerical with a pupp

In [22]:
type(test_result)

list

In [23]:
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
 

tf_idf_vec = TfidfVectorizer(use_idf=True,smooth_idf=False,ngram_range=(1,1),stop_words='english')

tf_idf_data = tf_idf_vec.fit_transform(test_result)
 



In [24]:
tf_idf_dataframe=pd.DataFrame(tf_idf_data.toarray(),columns=tf_idf_vec.get_feature_names())
print(tf_idf_dataframe)
print("\n")

    aching  adidas  administration  ads  aha  ahead  aig  allen  alpha  alto  \
0      0.0     0.0             0.0  0.0  0.0    0.0  0.0    0.0    0.0   0.0   
1      0.0     0.0             0.0  0.0  0.0    0.0  0.0    0.0    0.0   0.0   
2      0.0     0.0             0.0  0.0  0.0    0.0  0.0    0.0    0.0   0.0   
3      0.0     0.0             0.0  0.0  0.0    0.0  0.0    0.0    0.0   0.0   
4      0.0     0.0             0.0  0.0  0.0    0.0  0.0    0.0    0.0   0.0   
..     ...     ...             ...  ...  ...    ...  ...    ...    ...   ...   
95     0.0     0.0             0.0  0.0  0.0    0.0  0.0    0.0    0.0   0.0   
96     0.0     0.0             0.0  0.0  0.0    0.0  0.0    0.0    0.0   0.0   
97     0.0     0.0             0.0  0.0  0.0    0.0  0.0    0.0    0.0   0.0   
98     0.0     0.0             0.0  0.0  0.0    0.0  0.0    0.0    0.0   0.0   
99     0.0     0.0             0.0  0.0  0.0    0.0  0.0    0.0    0.0   0.0   

    ...      work  worked  world  would

7. Commenting on the matrix values

a matrix is created
where the rows are tf_idf values with respect to a nonstop english word in a document.
the coloums are non stop english words given in the entire file
where, 
tf= number of times a word occured in a document / total number of words in a document
idf = 1+ log(total number of documents in the dataset/total number of documents in which the nth word occur)
TF-IDF = TF*IDF
